lucene/core/src/java/org/apache/lucene/search/TopScoreDocCollector.java

```diff
@@ -207,16 +210,23 @@ public static TopScoreDocCollector create(int numHits, ScoreDoc after, int total
     }
   }
 
-  final int totalHitsThreshold;
   ScoreDoc pqTop;
+  final HitsThresholdChecker hitsThresholdChecker;
 
   // prevents instantiation
-  TopScoreDocCollector(int numHits, int totalHitsThreshold) {
+  TopScoreDocCollector(int numHits, int totalHitsThreshold, HitsThresholdChecker hitsThresholdChecker) {
     super(new HitQueue(numHits, true));
-    this.totalHitsThreshold = totalHitsThreshold;
+    assert hitsThresholdChecker != null;
+
     // HitQueue implements getSentinelObject to return a ScoreDoc, so we know
     // that at this point top() is already initialized.
     pqTop = pq.top();
+    this.hitsThresholdChecker = hitsThresholdChecker;
+  }
+
+  // Same as above but uses a local hits checker for hits threshold checks
+  TopScoreDocCollector(int numHits, int totalHitsThreshold) {
```

Comment:
Same as `TopFieldCollector`, we should have a single constructor that takes a `HitsThresholdChecker` and build the `totalHitsThreshold` logic in the static function.